### PR TITLE
Allow subscription query for sync events

### PR DESCRIPTION
**Pull request by 2can**
I want to merge this change because...

- it changes static event-object map to dynamically build one based on `WebhookEventTypeAsyncEnum` and `WebhookEventTypeSyncEnum`
- it enables the subscription query editor for synchronous events
- it shows the alert that dry run is yet not available for sync events

**PR intended to be tested with API branch:** 

### Screenshots

image

image

### Pull Request Checklist

1. [x] This code contains UI changes
2. [x] All visible strings are translated with proper context including data-formatting
3. [x] Attributes `[data-test-id]` are added for new elements
4. [x] Changes are mentioned in the changelog
5. [x] The changes are tested in different browsers and in light/dark mode

### Acceptance Criteria

1. [ ] This feature allows to define a synchronous subscription query of any shape
2. [ ] This feature must inform the user that Dry Run is not available for synchronous events

### Test environment config

API_URI=https://automation-dashboard.staging.saleor.cloud/graphql/
MARKETPLACE_URL=https://apps.saleor.io
APPS_MARKETPLACE_API_URI=https://marketplace-gray.vercel.app/api/v2/saleor-apps
SALEOR_APPS_ENDPOINT=https://apps.saleor.io/api/saleor-apps

### Do you want to run more stable tests?

To run all tests, just select the stable checkbox. To speed up tests, increase the number of containers. Tests will be re-run only when the "run e2e" label is added.

1. [ ] stable
2. [ ] giftCard
3. [ ] category
4. [ ] collection
5. [ ] attribute
7. [ ] productType
8. [ ] shipping
9. [ ] customer
10. [ ] permissions
11. [ ] menuNavigation
12. [ ] pages
13. [ ] sales
14. [ ] vouchers
15. [ ] homePage
16. [ ] login
17. [ ] orders
18. [ ] products
19. [x] app
20. [ ] plugins
21. [ ] translations
22. [ ] navigation
23. [ ] variants
24. [ ] payments

CONTAINERS=1

### CODE REVIEW

**Comment by szczecha** [comment]
Unexpected error when trying to select object not available for the dry run 

https://user-images.githubusercontent.com/30683248/217037202-46bba9ad-5cc0-417b-8404-b4963c62b462.mov

File: src/components/DryRun/DryRun.tsx [hunk 1]
```diff
@@ -33,15 +31,16 @@ interface DryRunProps {
...
+  data: WebhookFormData;
 }
 
-export const DryRun: React.FC<DryRunProps> = ({
+const DryRun: React.FC<DryRunProps> = ({
   setResult,
   showDialog,
   setShowDialog,
   query,
+  data: { syncEvents },
```
Comment (poulch) [inline src/components/DryRun/DryRun.tsx:43]: question: why do you pass hole data even if that component uses only syncEvents ?

File: src/custom-apps/components/WebhookEvents/WebhookEvents.tsx [hunk 2]
```diff
@@ -180,59 +187,52 @@ export default WebhookEvents;
...
+    object: string,
+    objectLength: number,
+  ) => {
+    const events = webhookTypes[object] || [];
+    const event = keywords.slice(objectLength).join("_");
+    events.push(!!event.length ? event : object);
+    webhookTypes[object] = events;
+  };
+
+  webhookEvents.map(key => {
```
Comment (poulch) [inline src/custom-apps/components/WebhookEvents/WebhookEvents.tsx:205]: suggestion: map return a new array which should be assigned to variable, if you don't want to assign the result of map to a new variable probably you should use forEach

File: src/custom-apps/components/WebhookEvents/WebhookEvents.tsx [hunk 3]
```diff
@@ -180,59 +187,52 @@ export default WebhookEvents;
...
+    const keywords = key.split("_");
+
+    // handle 2 word exceptions
+    if (multiWords.includes(keywords.slice(0, 2).join("_"))) {
+      setObject(keywords, keywords.slice(0, 2).join("_"), 2);
```
Comment (poulch) [inline src/custom-apps/components/WebhookEvents/WebhookEvents.tsx:210]: suggestion: `keywords.slice(0, 2).join("_")` is duplicated

File: src/custom-apps/components/WebhookEvents/WebhookEvents.tsx [hunk 4]
```diff
@@ -180,59 +187,52 @@ export default WebhookEvents;
...
 };
 
-const getEventName = (object: string, event: string) =>
-  [object, event].join("_").toUpperCase() as WebhookEventTypeSyncEnum;
+const getEventName = (object: string, event: string) => {
+  if (object === event) {
+    return object.toUpperCase() as WebhookEventTypeSyncEnum;
+  }
+  return [object, event].join("_").toUpperCase() as WebhookEventTypeSyncEnum;
+};
```
Comment (poulch) [inline src/custom-apps/components/WebhookEvents/WebhookEvents.tsx:238]: suggestion: move this code to separate file

File: src/components/DryRun/DryRun.tsx (hunk 1 above)
Comment (2can) [reply src/components/DryRun/DryRun.tsx]: changed

File: src/custom-apps/components/WebhookEvents/WebhookEvents.tsx (hunk 2 above)
Comment (2can) [reply src/custom-apps/components/WebhookEvents/WebhookEvents.tsx]: changed

File: src/custom-apps/components/WebhookEvents/WebhookEvents.tsx (hunk 3 above)
Comment (2can) [reply src/custom-apps/components/WebhookEvents/WebhookEvents.tsx]: fixed

File: src/custom-apps/components/WebhookEvents/WebhookEvents.tsx (hunk 4 above)
Comment (2can) [reply src/custom-apps/components/WebhookEvents/WebhookEvents.tsx]: done

**Comment by 2can** [comment]
@szczecha fix pushed

**Comment by szczecha** [comment]
Is there any task for adding a Preview label for some of the events?
Screenshot 2023-02-15 at 09 32 07

 
There are 3 events unknown type AnyEvent, Observability, NotifyUser
Screenshot 2023-02-15 at 09 25 02


And there is something odd with CustommerDeleted I checked the events list and looks like there is no such event
Screenshot 2023-02-15 at 09 29 45

Screenshot 2023-02-15 at 09 34 21

File: src/components/DryRun/utils.ts [hunk 5]
```diff
@@ -32,28 +32,34 @@ const getEventsFromQuery = (query: string) => {
...
+      .join("_")
+      .toUpperCase();
+
+    if (checkEventPresence(formattedEvent)) {
+      unavailableObjects.push(event);
+    }
+  });
+
+  return unavailableObjects;
+};
```
Comment (poulch) [inline src/components/DryRun/utils.ts:51]:
```suggestion
return   queryEvents.reduce((acc, event) => {
    const formattedEvent = event
      .split(/(?=[A-Z])/)
      .join("_")
      .toUpperCase();

    if (checkEventPresence(formattedEvent)) {
      acc.push(event);
    }
    
    return acc
  });
};
```

**Review by poulch — DISMISSED** [review]

File: src/components/DryRun/utils.ts (hunk 5 above)
Comment (2can) [reply src/components/DryRun/utils.ts]: Changed

**Comment by 2can** [comment]
@szczecha regarding your comment 2 issues have been created for that
- https://github.com/saleor/saleor-dashboard/issues/3199
- https://github.com/saleor/saleor-dashboard/issues/3200

**Review by poulch — APPROVED** [review]

File: src/custom-apps/components/WebhookEvents/utils.ts [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+    if (multiWords.includes(multiKeyword)) {
+      setObject(keywords, multiKeyword, 2);
+      return;
+    }
+
+    setObject(keywords, keywords[0], 1);
+  });
+
+  return webhookTypes;
+};
```
[the remark this answers is not in the document]
Comment (2can) [reply src/custom-apps/components/WebhookEvents/utils.ts]: changed

**Review by poulch — APPROVED** [review]

**Review by poulch — APPROVED** [review]

**Review by szczecha — APPROVED** [review]

**Pull request merged**